### ISSUE TITLE
API: MultiIndex.get_level_values should have label only / level only mode?

### ISSUE THREAD

**Issue by sinhrks** [pydata/pandas#10461]
Currently, ``MultiIndex.get_level_values`` handles both names and levels, it return different levels depending on its name (like ``.ix``). Adding an option (or separate method) like ``.loc`` and ``.iloc`` to avoid any unexpected results?

```
idx = pd.MultiIndex.from_tuples([(1, 'A'), (2, 'B')])
idx.get_level_values(0)
# Int64Index([1, 2], dtype='int64')
 
idx = idx.set_names([1, 0])
idx.get_level_values(0)
# Index([u'A', u'B'], dtype='object', name=0)
```

**Comment by shoyer** [comment]
Maybe `idx.get_level_values` and `idx.get_ilevel_values`? Or have we already locked in the mixed functionality of `get_level_values`?

**Comment by sinhrks** [comment]
I think we can't change ``get_level_values``. 

I think users prefer ``levels`` to ``names`` to avoid ambiguous, cos names can be duplicated or ``None``. Thus option 1 can cover most cases.

1. Add ``get_ilevel_values`` which accepts ``level`` only. No additional for ``names``.
2. Add ``get_ilevel_values`` which accepts ``level`` only and ``get_llevel_values`` (?) which accepts ``names`` only.
3. Add ``allow_level`` (?) and ``allow_names`` (?) options to ``get_level_values`` with default ``True``.

**Comment by jreback** [comment]
already there is inference to figure out if they mean positional (if no integers are in the names) or label based - eg like ix

but I have never seen this come up as a practical issue

I guess get_ilevel_values would be fine

ironically in .query we allow 

ilevel as a search term so this would be consistent at least

**Comment by jreback** [comment]
http://pandas.pydata.org/pandas-docs/stable/indexing.html#multiindex-query-syntax

**Comment by jreback** [comment]
just throwing it out there - could allow this:

df.index.ix(0) #current
df.index.loc('foo) #label only
df.index.iloc(0) #positional only

or maybe

df.index.ix[0]

further this might be useful

df.index.ix[0,1] -> grab first 2 levels

**Comment by sinhrks** [comment]
Right. Adding ``ix`` stacks sounds consistent and reasonable. Let me try.

**Comment by jorisvandenbossche** [comment]
I am -1 on adding `ix` like above. If we add `ix` (or loc/iloc to the index), I think the usage should be equivalent to as if you would have a column with the name 'index' and you are indexing `frame.index`. If it has other typical behaviour, we should just call it differently I think.

**Comment by Dr-Irv** [comment]
This relates to #12223 and #15262 which are bugs due to allowing index names to be integers.  For those, the issue is the _internal_ use of `get_level_values()` and the ambiguity of the argument if one passes an integer, and some of the names of the index are integers.  I have two proposals that could fix this:
1. Modify the signature of `get_level_values()` to include an optional argument `try_names_first=True` that preserves current behavior.  Modify _internal_ calls in pandas that use `get_level_values()` to pass `try_names_first=False` when the caller knows it is passing the level number instead of a name.
2. Create a new _internal_ method called `_get_level_values_by_level_number()` that is not in the published API, but is used internally in pandas code. Modify _internal_ calls in pandas that use `get_level_values()` to call `get_level_values_by_level_number()`  when the caller knows it is passing the level number instead of a name.

I think I am in favor of (2) as I actually think it gives a very small performance boost, but would like to hear the opinions of @jreback, @jorisvandenbossche and @sinhrks before I embark on making those changes.

**Comment by jreback** [comment]
2) is fine
